tests/test_ote_training.py

```diff
@@ -0,0 +1,617 @@
+        labels_list = self.labels_schema.get_labels(False)
+        self.dataset.set_project_labels(labels_list)
+        print(f'train dataset: {len(self.dataset.get_subset(Subset.TRAINING))} items')
+        print(f'validation dataset: {len(self.dataset.get_subset(Subset.VALIDATION))} items')
+
+        logger.debug('Load model template')
+        self.model_template = parse_model_template(self.template_file_path)
+
+        hyper_parameters = self.model_template.hyper_parameters.data
+        set_values_as_default(hyper_parameters)
+
+        logger.debug('Setup environment')
+        params = create(hyper_parameters)
+        logger.debug('Set hyperparameters')
+        params.learning_parameters.num_iters = self.num_training_iters
+        if self.num_training_iters < 20:
+            num_checkpoints = 2
+        elif self.num_training_iters < 1000:
+            num_checkpoints = 10
+        else:
+            num_checkpoints = 30
+
+        #### TODO: fixed parameter, delete this
+        params.learning_parameters.batch_size = 2
+
+        params.learning_parameters.num_checkpoints = num_checkpoints
+
+        self.environment, self.task = self._create_environment_and_task(params,
+                                                                        self.labels_schema,
+                                                                        self.model_template)
+
+        logger.debug('Train model')
+        self.output_model = ModelEntity(
+            self.dataset,
+            self.environment.get_model_configuration(),
+            model_status=ModelStatus.NOT_READY)
+
+        self.copy_hyperparams = deepcopy(self.task._hyperparams)
+
+        self.task.train(self.dataset, self.output_model)
+        logger.info(f'performance={self.output_model.performance}')
+
+        # To check that the training performance is valid
+        # and may be converted to score name and value
+        self.get_training_performance_as_score_name_value()
+
+    def get_training_performance_as_score_name_value(self):
+        training_performance = getattr(self.output_model, 'performance', None)
+        if training_performance is None:
+            raise RuntimeError('Cannot get training performance')
+        return performance_to_score_name_value(training_performance)
+
+    def run_ote_training_once(self, data_collector):
+        if self.was_training_run and self.stored_exception_training:
+            logger.warn('In function run_ote_training_once: found that previous call of the function '
+                        'caused exception -- re-raising it')
+            raise self.stored_exception_training
+
+        if not self.was_training_run:
+            try:
+                self._run_ote_training()
+                self.was_training_run = True
+            except Exception as e:
+                self.stored_exception_training = e
+                self.was_training_run = True
+                raise e
```

Comment:
`self.was_training_run` is set to `True` either way.

Reply:
Yes, it was for the rare case if somebody will insert checking of the flag into `_run_ote_training` or one of the functions called from it. 
We cannot set it before `_run_ote_training`, and cannot insert it in one place only, since storying of exception will require setting this flag too, end the code re-raises the exception just after it.